aws-janitor/resources/s3_bucket.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type S3Bucket struct{}
+
+func (S3Bucket) MarkAndSweep(opts Options, set *Set) error {
+	logger := logrus.WithField("options", opts)
+	if !opts.EnableS3BucketClean {
+		logger.Info("Disable s3 bucket clean")
+		return nil
+	}
+	svc := s3.New(opts.Session, &aws.Config{Region: aws.String(opts.Region)})
+
+	// ListBucket lists all buckets owned by the authenticated sender of the request
+	// https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3#S3.ListBuckets
+	// regardless the region configured in client.
+	resp, err := svc.ListBuckets(&s3.ListBucketsInput{})
+	if err != nil {
+		return err
+	}
+	for _, b := range resp.Buckets {
+		bucketName := *b.Name
+
+		region, err := s3manager.GetBucketRegion(context.Background(), opts.Session, bucketName, opts.Region)
+        if err != nil || region != opts.Region {
+			continue
+		}
+		respTags, err := svc.GetBucketTagging(&s3.GetBucketTaggingInput{Bucket: aws.String(bucketName)})
+		if err != nil {
+			continue
+		}
+		tags := fromS3Tags(respTags.TagSet)
+		if !set.Mark(opts, s3Bucket{bucketName: bucketName}, nil, tags) {
+			continue
+		}
+		var objectsToDelete []*string
+		// Before deleting the bucket, delete its objects.
+		if err := svc.ListObjectsV2Pages(&s3.ListObjectsV2Input{Bucket: aws.String(bucketName)},
+			func(page *s3.ListObjectsV2Output, _ bool) bool {
+				for _, object := range page.Contents {
+					logger.Warningf("Deleting object: %s at bucket: %s", *object.Key, bucketName)
+					objectsToDelete = append(objectsToDelete, object.Key)
+				}
+				return true
+			}); err != nil {
+				logger.Warningf("ListObjectsV2Pages failed for bucket: %s : %v", bucketName, err)
+				continue
+			}
+		// If the bucket stores janitor state objects, not to delete it.
+		isStateBucket := false
+		for _, objectKey := range objectsToDelete{
+			if isStateBucket {
+				break
+			}
+			resp, err := svc.GetObjectTagging(&s3.GetObjectTaggingInput{Bucket: aws.String(bucketName), Key: aws.String(*objectKey)})
```

Comment:
Feels a little bit too complex and heavy weight, also buckets can be arbitrarily large... Also if it's a new bucket protection probably would work anyway. I guess we can assume state buckets are properly tagged with do not delete tags then and add warning to the flag doc.